// Copyright 2025 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option cc_enable_arenas = true;

package ray.rpc;

// Represents a sub-operator within a main operator
message SubOperator {
  // Name of the sub-operator
  string name = 1;

  // Unique identifier for the sub-operator
  string id = 2;
}

// Represents a data processing operator in the DAG
message Operator {
  // Name of the operator
  string name = 1;

  // Unique identifier for the operator
  string id = 2;

  // List of operator IDs that this operator depends on for input
  repeated string input_dependencies = 3;

  // List of sub-operators contained within this operator
  repeated SubOperator sub_operators = 4;
}

// Represents the complete structure of the operator DAG
message OperatorDAG {
  // List of all operators in the DAG
  repeated Operator operators = 1;
}

// Top-level message containing full metadata about a Ray Data execution
message ExportDataMetadata {
  // The operator DAG structure
  OperatorDAG dag = 1;
}
